src/main/kotlin/com/dprint/listeners/OptimizeJavaImports.kt

```diff
@@ -0,0 +1,83 @@
+package com.dprint.listeners
+
+import com.dprint.config.ProjectConfiguration
+import com.dprint.core.Bundle
+import com.dprint.core.LogUtils
+import com.dprint.services.editorservice.EditorServiceManager
+import com.dprint.services.editorservice.FormatResult
+import com.intellij.lang.ImportOptimizer
+import com.intellij.lang.ImportOptimizer.CollectingInfoRunnable
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiJavaFile
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+private val LOGGER = logger<OptimizeJavaImports>()
+private const val TIMEOUT = 10L
+
+/**
+ * This listener formats Java files when optimize imports called.
+ */
+class OptimizeJavaImports : ImportOptimizer {
+
+    override fun supports(file: PsiFile): Boolean {
+        if (file !is PsiJavaFile) return false
+
+        val project = file.project
+        if (!project.service<ProjectConfiguration>().state.enabled) return false
+
+        val documentManager = PsiDocumentManager.getInstance(project)
+        val document = documentManager.getDocument(file) ?: return false
+
+        val editorServiceManager = project.service<EditorServiceManager>()
+        val fileManager = FileDocumentManager.getInstance()
+        return fileManager.getFile(document)?.path?.let { editorServiceManager.canFormatCached(it) } ?: false
+    }
+
+    override fun processFile(file: PsiFile): Runnable {
+        val project = file.project
+        val manager = FileDocumentManager.getInstance()
+
+        val documentManager = PsiDocumentManager.getInstance(project)
+        val document = documentManager.getDocument(file) ?: return Runnable { }
+        val editorService = project.service<EditorServiceManager>()
+
+        var error: String? = null
+        var formattedContent: String? = null
+        lateinit var future: CompletableFuture<Unit>
+        manager.getFile(document)?.let { virtualFile ->
+            LogUtils.info(Bundle.message("optimize.imports.run", virtualFile.path), project, LOGGER)
+            val formatHandler: (FormatResult) -> Unit = { formatResult ->
+                formatResult.error?.let {
+                    error = it
+                    LogUtils.warn(Bundle.message("optimize.imports.error", it), project, LOGGER)
+                } ?: run {
+                    formattedContent = formatResult.formattedContent
+                }
+            }
+
+            future = editorService.format(virtualFile.path, document.text, formatHandler)
+        }
+        future.get(TIMEOUT, TimeUnit.SECONDS)
```

Comment:
Should we be calling get on the future here? I am pretty sure this will block the EDT and if so we should probably push it down into the `run` method of our runnable.

Reply:
Hmmmm this doc from the `processFile` method seems to indicate it is will still block the UI as it is run in the swing thread so either way we will be blocking the UI.

```
 /**
   * Implementers of the method are expected to perform all necessary calculations synchronously and return a {@code Runnable},
   * which performs modifications based on preprocessing results.
   * processFile() is guaranteed to run with {@link com.intellij.openapi.application.Application#runReadAction(Runnable)} privileges and
   * the Runnable returned is guaranteed to run with {@link com.intellij.openapi.application.Application#runWriteAction(Runnable)} privileges.
   *
   * One can theoretically delay all the calculation until Runnable is called but this code will be executed in Swing thread thus
   * lengthy calculations may block user interface for some significant time.
   *
   * @param file to optimize an imports in. It's guaranteed to have a language this {@code ImportOptimizer} have been
   * issued from.
   * @return a {@code java.lang.Runnable} object, which being called will replace original file imports with optimized version.
   */
  @NotNull
  Runnable processFile(@NotNull PsiFile file);
```

Did you get any EDT warnings while running this at any stage?

Reply:
Yes, I did. I have not found any way to run long task as a reaction on pre-commit check/action. Was also looking at another pre-commit actions, including:
- intellij internal reformat action
- intellij pre-commit code analyze action
- prettier plugin implementation
all of them at some point are doing blocking actions inside EDT.
🤷‍♂️ 
I hope I'll find something better but could not so far.